consensus/istanbul/proxy/proxy_handler.go

```diff
@@ -0,0 +1,352 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package proxy
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// This struct defines the handler that will manage all of the proxies and
+// validator assignments to them
+type proxyHandler struct {
```

Comment:
I see that `proxyEngine` access directly `proxyHandler` fields. In particular; all the channels.

IMO, we should avoid that, and instead define methods for all those actions. That way we define a cleaner interface to the `proxyHandler`. 

As a rule of thumb, nobody should be accessing "private" fields. Something not enforced within the same package in golang.

Using a channel is an implementation detail, should not be something `proxyEngine` is aware of.